In [7]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings(action='ignore')

from matplotlib import gridspec
%matplotlib inline


from math import factorial

from scipy import stats
from scipy.stats import randint                 # randint is used to generate random variables following a uniform distribution.
from scipy.stats import skewnorm                # variables having a skewed normal distribution.

import random

In [155]:
plt.rcParams['figure.figsize'] = [15,8]         # setting up the figure sizes

# DAY 1 

 
## The study of statistics is mainly divided into 2 parts:
    
####    1. Descriptive stats.
####    2. Inferential Stats.
    


# Descriptive Stats:
    
#### 1. Measure of Central tendency:
        
       -- Mean.
       -- Median.
       -- Mode.
        
####    2. Measure of Dispersion:
        
       -- Range.
       -- Variance.
       -- Standard Deviation.
       -- Coeffiecent of Variation.
       -- Interquartile Range(IQR)
        
####    3. Skewness  &  Kurtosis.
            
####   4. Covariance  &  Correlation.
    
    

# Inferential Stats:

#  Probability Distributions.

### ----------- Discrete Probability Distributions -----------
  #### 1. Discrete Uniform Distribution

In [158]:
# To find P(x=3 or x=4)
# Use random.pmf 
# low is minimum of x  
# high is maximum of x

prob = randint.pmf(k=3, low=1, high=7) + randint.pmf(k=4, low=1, high=7)
print('The probabilty that a machine chosen at random is either 3 or 4 is: ',round(prob,4))


The probabilty that a machine chosen at random is either 3 or 4 is:  0.3333


#### 2. Discrete Bernoulli Distribution.

In [159]:
# Bernoulli distribution has only two possible outcomes:
# From the above question theres either goal or not a goal,So bernoulli    
# use stats.bernoulli.pmf( k=(), p=())

# To find P(X=1)
# p = 0.7 given.

p = 0.7
prob = stats.bernoulli.pmf(k=1,p=0.7)
print('The probability that player scores a goal for the next free kick is:',round(prob,4))

The probability that player scores a goal for the next free kick is: 0.7


#### 3. Discrete Binomial Distribution.

In [160]:
# Binomial distribution is generated when we repeat a bernoulli experiment n number of times.

# The question has a possibility either a customer complains or not which is two possible outcomes(bernoulli) repeated n times 
# Hence Binomial distribution.
# use stats.binom.pmf(k=(), n=(), p=())
# n=20 number of customers
# p=0.25, 25% probabilty
# P(X=3)

prob = stats.binom.pmf(k=3, n=20, p=0.25)
print('The probability that exactly 3 customers will complain about the purchased products is: ',round(prob,4))


The probability that exactly 3 customers will complain about the purchased products is:  0.1339


In [162]:
# When probability more than like that,
# use sf() - survival function.
# stats.binom.sf()

prob = stats.binom.sf(k=3, n=20, p=0.25)
print('The probability that more than 3 customers will complain about the purchased products is: ',round(prob,4))

The probability that more than 3 customers will complain about the purchased products is:  0.7748


In [165]:
# its binomal distribution.
# use stats.binom.pmf()

prob = stats.binom.pmf(k=35, n=50, p=11/15)
print('The probability that out of 50 shots fired, exactly 35 will hit the target is: ',round(prob,4))

The probability that out of 50 shots fired, exactly 35 will hit the target is:  0.1065


#### 4. Discrete Poisson Distribution.

In [166]:
# use 'poisson.sf()' to calculate the value of survival function (1 - cdf). i.e. P(X > x)
# calculate the probability that more than 70 pizzas will be sold 
# pass the required value of pizzas to the parameter, 'k' 
# pass the average number of pizzas to the parameter, 'mu'

prob = stats.poisson.sf(k = 70, mu = 67)
print('The probability that the number of pizza exceeds 70 in a day is:',round(prob,4))

The probability that the number of pizza exceeds 70 in a day is: 0.3285


### ---------- Continous Probability Distributions ----------

#### 1. Continous Uniform distribution.

In [168]:
# The distance b/w gas supplying center and the city 200 kms.
# to find: P(50<=X<=150), where pipe leaks in the middle of 100 kms.
# its continous uniform distribution over the interval (0,200)

# first will find the cdf of both 150 and 50.
# loc is the start point of the interval of the parameter.
# scale is the difference b/w the end and start point of interval of the parameter.
prob_150 = stats.uniform.cdf(x=150, loc=0, scale=200)
prob_50 = stats.uniform.cdf(x=50, loc=0, scale=200)

# now by the difference of the above we get the required probability:
req_prob = prob_150 - prob_50
print('The probability that the pipe leaks on the middle 100 kms is:',round(req_prob,4))


The probability that the pipe leaks on the middle 100 kms is: 0.5


#### 2. Continous Normal Distribution.

In [171]:
# to find P(X<100)

# average and std of the data.
avg = 300
std = 127

# standardize the variable with x=100,  (z=(x-avg)/std)
X = 100
z = (X-avg)/std

prob = stats.norm.cdf(z)
print('The probability that the users are spending less than or equal to 100 minutes per day is:',round(prob,2))


The probability that the users are spending less than or equal to 100 minutes per day is: 0.06


In [174]:
# to find P(X>400) , greater so, use sf()

# average and std of the data.
avg = 300
std = 127

# standardize the variable with x=400,  (z=(x-avg)/std)
X = 400
z = (X-avg)/std

prob = stats.norm.sf(z)
print('The probability that people are spending more than 400 minutes per day is:',round(prob,4))


The probability that people are spending more than 400 minutes per day is: 0.2155


In [177]:
# To find P(X) b/w 250 and 350

# average and std of the data.
avg = 300
std = 127

# Standardize the variable with x1=250 and x2=350
x1=250
x2=350
z1 = (x1-avg)/std
z2 = (x2-avg)/std

p1 = stats.norm.cdf(z1)
p2 = stats.norm.cdf(z2)

prob = p2 - p1
print('The reqiured probability is:', round(prob,2))

The reqiured probability is: 0.31


In [178]:
# To find P(X>513.5), greater use sf()

# average and std of the data.
avg = 500
# std not given directly variance is given So,
var = 100
std = np.sqrt(var)

# standardize the variable with x=513.5
X = 513.5
z = (X-avg)/std

prob = stats.norm.sf(z)
print("The required probability is:", round(prob,4))


The required probability is: 0.0885


# Sampling

#### Types of Sampling:
     
     - Simple random Sample.
     - Stratified.
     - Systematic.
     - Cluster.

#### 1. Simple Random Sample.

In [183]:
# Given population 
data = [18, 13, 16, 17, 15, 21, 19, 12]
print('population:',data)

# get sample with replacement.
# use random.choices

sample = random.choices(population=data, k=20)
print('Sample:',sample)

population: [18, 13, 16, 17, 15, 21, 19, 12]
Sample: [13, 15, 21, 21, 15, 19, 17, 19, 12, 17, 19, 21, 12, 18, 12, 19, 12, 12, 16, 21]


In [185]:
# Given population:
data = [21, 93, 62, 76, 73, 20, 56, 95, 41, 36, 38, 13, 80, 88, 34, 18, 40, 11, 25, 29, 61, 23, 82, 10, 92, 69, 60, 87, 14, 91]
print('population:',data)
# get sample without replacement.
# use random.sample

sample = random.sample(population=data, k=10)
print('Sample:',sample)

population: [21, 93, 62, 76, 73, 20, 56, 95, 41, 36, 38, 13, 80, 88, 34, 18, 40, 11, 25, 29, 61, 23, 82, 10, 92, 69, 60, 87, 14, 91]
Sample: [56, 88, 91, 76, 23, 36, 34, 82, 40, 14]


# CENTRAL LIMIT THEOREM :

- let x1,x2,x3,....xn be the random sample drawn from a population.

- with mean 𝜇 and std 𝜎 .

- The central limit theorem states that, for sufficiently large n, the sample mean 𝑋 follows an approximately normal distribution with mean 𝜇 and std $\frac{\sigma}{\sqrt{n}}$.

- i.e. $\frac{\overline{X} - \mu}{\frac{\sigma}{\sqrt{n}}}$ follows a standard normal distribution.

- For a normally distributed population, the above result holds for any sample size.

- For the population with distribution other than normal, generally, the sample size greater than or equal to 30 is considered as the large sample size.

#### let us plot some graphs to demonstrate the CLT for positive, negative and normal positive distribution.

# DAY 2

In [187]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings(action='ignore')

from matplotlib import gridspec

from math import factorial

from scipy import stats
from scipy.stats import randint
from scipy.stats import skewnorm

import random

import statistics

from statsmodels.stats import weightstats as stests    # to perform z-test

from scipy.stats import shapiro                        # to test normality

from statsmodels.stats import power                    # to calculate the power of test.

In [189]:
plt.rcParams['figure.figsize'] = [15,8]                # to set plot size

# Parameter Estimation.
- The value associated with the characteristic of the population is known as a parameter.
- and the characteristic of the sample is described by a statistic.
- Usually, the population parameters are not known in most of the real-life problems.
- Thus, we consider a subset of the population(sample) to estimate the population parameter using sample statistic.

## There are two methods of  to estimate the population parameter.
- Point Estimation.
- Interval Estimation.

### 1. Point Estimation.
###### Point Estimation for mean,
###### Point Estimation for propotion.
- This method considers a single value(sample statistic) as the population parameter.

In [190]:
# ---------- THIS IS POINT ESTIMATION for MEAN ---------- #

# Given population:
Grades = [59.1, 65.0, 75.8, 79.2, 95.0, 99.8, 89.1, 65.2, 41.9, 55.2, 94.8, 84.1, 83.2, 74.0, 75.5, 76.2, 79.1, 80.1, 
          92.1, 74.2, 59.2, 64.0, 75, 78.2, 95.6, 97.8, 89.5, 64.2, 41.8, 57.2, 85, 91.4, 81.8, 74.6, 90]

# To draw a sample for grade of 20 students without replacement.
# use random.seed for get the same sample each time u run the code.
random.seed(1)

# get the sample from population
samp_grades = random.sample(population=grades, k=20)

# the point estimate of the population mean is equal to sample mean.
# so, get the sample mean.
samp_mean = np.mean(samp_grades)

print('The point estimate for the population mean is',samp_mean)

The point estimate for the population mean is 75.32500000000002


In [192]:
# ---------- THIS IS POINT ESTIMATION for PROPOTION ---------- #

# Total count of portfolios.
N = 50

# number of portfolios in sample.
n = 13

# number of underproforming portfolois in a sample.
x = 8

# sample propotion.
p_samp = x/n

# estimate the number of underperforming portfolios.
num_port = p_samp*N

print('The number of underperforming portfolios are:',round(num_port))

The number of underperforming portfolios are: 31


### 2. Interval Estimation .
###### Interval estimation for mean.
###### Interval estimation for propotion.
- This method considers the range of values in which the population parameter is likely to lie.
- The confidence interval is an interval that describes the range of values in which the parameter lies with a specific probability. 
- The confidence interval for the population mean with $100(1-\alpha)$% confidence level is given as: $\overline{X} \pm Z_{\frac{\alpha}{2}}\frac{\sigma}{\sqrt{n}}$
- Where, <br>
$\overline{X}$: Sample mean<br>
$\alpha$: Level of significance<br>
$\sigma$: Population standard deviation<br>
$n$: Sample size

#####  i) Large Sample size.
##### ii) Small Sample size.

### Interval Estimation for mean.

#### 1. Large Sample size n>30.

In [216]:
# let us find the Z-values for the different alpha values.

# create an empty dataframe to store the aplha values and coressponding Z-values.
df = pd.DataFrame()

# create a dictionary of different aplha values.
alpha = [0.1, 0.05, 0.02, 0.01]

# use for loop to calculate the z-value of each aplha value.
for i in range(len(alpha)):
    alpha_by_2 = alpha[i] / 2
    
# use stats.norm.isf() to find tge z-values\.
# pass the value of alpha_by_2 to the parameter 'q'
    Z = np.abs(round(stats.norm.isf(q = alpha_by_2),4))

# create a dataframe using dictionary to store the alpha and corresponding Z-value
    row = pd.DataFrame({"alpha":alpha[i],"Z_alpha_by_2":Z},index = [i])

# append the row to the dataframe df
    df = df.append(row)

df


alpha  Z_alpha_by_2
0   0.10        1.6449
1   0.05        1.9600
2   0.02        2.3263
3   0.01        2.5758

In [6]:
# Given Sample:
Weight = [59.1, 65.0, 75.8, 79.2, 95.0, 99.8, 89.1, 65.3, 41.9, 55.2, 94.8, 84.1, 83.2, 74.0, 75.5, 76.2, 79.1, 80.1, 
         92.1, 74.2, 59.2, 64.0, 75, 78.2, 95.6, 97.8, 89.5, 64.2, 41.8, 57.2, 85, 91.4, 81.8, 74.6, 90]

# number of observations of the sample.
n = len(Weight)

# Given std of the population is 8 kg.
std_pop = 8

# To calculate the 90% confidence interval.
# pass the sample mean to the parameter, 'loc'.
# pass the scaling factor (std_pop/n^(1/2)) to the parameter, 'scale'.
interval = stats.norm.interval(0.90, loc=np.mean(weight), scale=(std_pop/np.sqrt(n)))

print('The 90% confidence interval of population mean is :',np.round(interval,2))

The 90% confidence interval of population mean is : [74.46 78.91]


### Let us calculate the confidence interval without using the built-in function.

In [5]:
# given sample
weight = [59.1, 65.0, 75.8, 79.2, 95.0, 99.8, 89.1, 65.2, 41.9, 55.2, 94.8, 84.1, 83.2, 74.0, 75.5, 76.2, 79.1, 80.1, 
          92.1, 74.2, 59.2, 64.0, 75, 78.2, 95.6, 97.8, 89.5, 64.2, 41.8, 57.2, 85, 91.4, 81.8, 74.6, 90]

# calculate sample mean
avg  = np.mean(weight) 

# given population standard deviation 
std_pop = 8

# number of observations in sample
n = len(weight)

# Z-value for 90% confidence level                
z_alpha_by_2 = 1.6449

# calculate the lower and upper bound of interval using the formula
lower_int = avg - (z_alpha_by_2 * (std_pop / np.sqrt(n)))
upper_int = avg + (z_alpha_by_2 * (std_pop / np.sqrt(n)))

# print the 90% confidence interval
# use 'round()' to round-off the value to 2 digits
print('90% confidence interval for population mean is', (round(lower_int,2), round(upper_int,2)))

90% confidence interval for population mean is (74.46, 78.91)


In [226]:
# number of the observations from the sample.
n = 40

# sample mean.
samp_mean = 182

# sample std.
sample_std = 30

# to calculate the 95% confidence interval.
interval = stats.norm.interval(0.95, loc=samp_mean, scale=(sample_std/np.sqrt(n)))

print('The 95% confidence interval for the pop mean is:',np.round(interval,2))

The 95% confidence interval for the pop mean is: [172.7 191.3]


### 2. Small Sample Size.

### Interval Estimation For propotion.

In [228]:
# total count of portfolios
N = 50

# number of portfolios in a sample
n = 13

# number of underperforming portfolios in a sample
x = 8

# sample proportion
p_samp = x/n

# calculate the 99% confidence interval
# pass the sample proportion to the parameter, 'loc'
# pass the scaling factor ((p_samp*(1-p_samp))/n))^0.5) to the parameter, 'scale'
interval = stats.norm.interval(0.99, loc = p_samp, scale = np.sqrt((p_samp*(1-p_samp))/n))

print('99% confidence interval for population proportion is', interval)

99% confidence interval for population proportion is (0.26782280814713805, 0.9629464226220927)


# Sampling Error.
- Sampling error is (Sample_mean - Population_mean)

In [196]:
# Given population and the sample.
data = [21, 93, 62, 76, 73, 20, 56, 95, 41, 36, 38, 13, 80, 88, 34, 18, 40, 11, 25, 29, 61, 23, 82, 10, 92, 69, 60, 87, 14, 91, 94, 49, 57, 83, 96, 55, 79, 52, 59, 39, 58, 17, 19, 98, 15, 54, 48, 46, 72, 45, 65, 28, 37, 30, 68, 75, 16, 33, 31, 99, 22, 51, 27, 67, 85, 47, 44, 77, 64, 97, 84, 42, 90, 70, 74, 89, 32, 26, 24, 12, 81, 53, 50, 35, 71, 63, 43, 86, 78, 66]
sample = [10, 22, 47, 66, 11, 57, 77, 98, 31, 63, 74, 84, 50, 96, 88, 92, 70, 54, 65, 44, 16, 72, 20, 90, 43]

# find the population mean and sample mean.
pop_mean = np.mean(data)
samp_mean = np.mean(sample)

sample_error = (pop_mean-samp_mean)
print('The sampling error for mean :',abs(sample_error))

The sampling error for mean : 3.1000000000000014


# Types of Hypothesis

- ***Null Hypothesis***: 
The null hypothesis is the claim suggesting 'no difference'. It is denoted as H<sub>0</sub>.

- ***Alternative Hypothesis***: It is hypothesis that is tested against the null hypotheisis.
- The acceptance or rejection of the hypothesis is based on the likelihood of H<sub>0</sub> being true.
- it is denoted by H<sub>a</sub> or  H<sub>1</sub>.


# Types of test
- **Two tailed Test** 

- **one tailed Test**

#### 1. Two tailed Test.
- If we test the population mean ($\mu$) with a specific value ($\mu_{0}$) the null hypothesis is: $H_{0}: \mu = \mu_{0}$. 
The alternative hypothesis for the two tailed test is given as: $H_{1}: \mu \neq \mu_{0}$

In [ ]:
# Null Hypothesis          : Ho : u = 2.65
# Alternative Hypothesis   : H1 : u ≠ 2.65

#### 2. One tailed Test
- One tailed test considers the value of the population parameter is less than or greater than (but not both) a specific value.

- If we test the population mean ( 𝜇 ) with a specific value ( 𝜇0 ) the null hypothesis is:  **𝐻0:𝜇≤𝜇0**  and the alternative hypothesis is  **𝐻1:𝜇>𝜇0** the one tailed test is also known as a **right-tailed test**.

- If we test the population mean ( 𝜇 ) with a specific value ( 𝜇0 ) the null hypothesis is:  **𝐻0:𝜇≥𝜇0**  and the alternative hypothesis is  **𝐻1:𝜇<𝜇0** the one tailed test is also known as a **left-tailed test**.

In [233]:
# Whenever there more than or less than first consider alternative hypothesis and then consider the null.
#So,

# Null Hypothesis        : Ho : Machine efficiency ≤ 8.
# Alternative Hypothesis : H1 : Machine efficiency > 8.


# This is an example for RIGHT TAILED TEST.

In [ ]:
# Null Hypothesis       : Ho : speed > 54    
# Alternative Hypotheis : H1 : speed < 54           # first consider the Alternative Hypothesis.

# This is an example for LEFT TAILED TEST.

# Hypothesis Tests with Z Statistic.
- Let us perform one sample Z test for the population mean. We compare the population mean with a specific value. The sample is assumed to be taken from a population following a normal distribution.
- **To check the normality of the data**, a test for normality is used. **The Shapiro-Wilk Test is one of the methods used to check the normality**. The hypothesis of the test is given as:

##### 1. Z - statistic test.

In [ ]:
# step 1: The null and alternative hypothesis.

# Null Hypothesis,           mileage : Ho : u = 25
# Alternative Hypothesis,    mileage : H1 : u ≠ 25
           # This is two tailed test.

In [238]:
# Step 2: Find the z-value for the alpha.

# Here = 0.01, for a two-tailed test calculate the critical z-value.
# calcualte the z-value for 99% of confidence level.
# pass the value alpha/2 to parameter q.

z_val = np.abs(round(stats.norm.isf(q=0.01/2),2))
print('The critical value for two tailed Z-test is:',z_val)

The critical value for two tailed Z-test is: 2.58


In [240]:
# Here if z-score is less than -2.58 or greater than 2.58 then we reject the null hypothesis.

# step 3: To calculate the Z-test Statistic.

# number of samples.
n = 45

# population mean.
pop_mean = 25

# sample mean.
samp_mean = 24

# standard deviation the data.
pop_std = 2.5

z_score = (samp_mean - pop_mean) / (pop_std / np.sqrt(n))
print('The Z-score is:',round(z_score,2))


The Z-score is: -2.68


##### 2. P-Value test.

In [242]:
# to calculate the p-value for the test stastistic.
# use 'cdf()' to calculate (Z <= z_score)
p_value = stats.norm.cdf(z_score)

# for a two-tailed test multiply the p-value by 2.
p_value = p_value*2
print('P-value:',p_value)

P-value: 0.007290358091535638


##### 3. Confidence Interval test.

In [246]:
# To calculate the 99% confidence interval for the pop mean.
n = 45
pop_mean = 25
samp_mean = 24
std = 2.5

interval = stats.norm.interval(0.99, loc=samp_mean, scale=std/np.sqrt(n))
print("Confidence Interval:",np.round(interval,2))

Confidence Interval: [23.04 24.96]


## **Conclusion**:

- Here the **z-score** is less than -2.58.
- **p-value** is less than 0.01.
- **confidence interval** do not contain the value in the null hypothesis(i.e, 25).
- Thus we reject the null hypothesis and conclude that there is enough evidence to claim mean mileage is different from 25 kmph.

In [ ]:
# FOR MORE UNDERSTANDING VIEW THE GRAPH IN FACULTY NOTEBOOK:

##### 1. Z- statistic test.

In [ ]:
# step1: The null and alternative Hypothesis.

# Null hypothesis         Ho : u = 82
# Alternative hypothesis  H1 : u ≠ 82

# This is Two tailed test.

In [249]:
# step2: Here ⍺ = 0.05, for two tailed test calculate the critical Z-value.

# calculate the z-value.
# pass the value 'alpha/2' for parameter two because two tailed test.
z_value = stats.norm.isf(q=0.05/2)
print('The crictical value for the two-tailed Z-test is:',round(z_value,2))

The crictical value for the two-tailed Z-test is: 1.96


In [251]:
# step3: To calculate the z-score.

n=40
pop_mean = 82
pop_std = 15
samp_mean = 95

z_score = (samp_mean - pop_mean) / (pop_std/np.sqrt(n))
print('z-score:',round(z_score,2))

z-score: 5.48


##### 2.P-value test.

In [252]:
# To calcualte the p-value
# use 'cdf()' to calculate P(Z <= -z_score)
p_value = stats.norm.cdf(-z_score)

# for a two-tailed test multiply the p-value by 2.
req_p = p_value*2
print('p-value:', req_p)

p-value: 4.222565249683579e-08


##### 3.Confidence Interval .

In [253]:
# to calculate the 95% confidence interval of the pop mean.
interval = stats.norm.interval(0.95, loc=samp_mean, scale=pop_std/np.sqrt(n))
print('Confidence Interval:',interval)

Confidence Interval: (90.35153725771579, 99.64846274228421)


# Conclusion.

- Here the z score is greater than 1.96.
- the p-value is less than 0.05.
- also the confidence interval do not contain the value in the null hypothesis (i.e. 82).
- thus we reject the null hypothesis and thus, there is enough evidence to conclude that the average calories in a slice bread are not 82.

In [256]:
acid_amt = [2.57, 3.06, 3.28 , 3.24, 2.79, 3.40, 3.36, 3.07, 2.46, 3.03, 3.05, 2.94, 3.46, 3.19, 3.09, 2.81, 3.13, 2.88, 
            2.76, 2.75, 3.17, 2.89, 2.54, 3.18, 3.08, 2.60, 3.06, 3.13, 3.11, 3.08, 2.93, 2.90, 3.06, 2.97, 3.24, 2.86, 
            2.87, 3.18, 3, 2.95]

##### 1. Z-Statistic Test.

In [257]:
# step1: The null and alternative Hypothesis.

# Null hypothesis         Ho : u > 3 
# Alternative hypothesis  H1 : u < 3 ,consider this first 

# This is left tailed test.

In [258]:
# step2: Here ⍺ = 0.05, for a one-tailed test calculate the critical z-value

# calculate the z-value for 95% confidence interval of pop mean.
# pass the value 'alpha' for the parameter q.(for one tailed test dont divide alpha by 2)
z_value = stats.norm.isf(q=0.05)
print('The critical value for one tailed Z-test is:',round(z_value,2))

The critical value for one tailed Z-test is: 1.64


In [259]:
# step3: To find Z-score.

n = 40 
pop_mean = 3

# This two values are not given:
samp_mean = np.mean(acid_amt)      # will get it by mean of given sample. 

# pop_std = np.std(acid_amt) # donot do this because,
pop_std = statistics.stdev(acid_amt)
#print(statistics.stdev(acid_amt)) #Why this was used we use sample std deviation
# use an unbiased estimate of sample variance. this has to be taken care when
#we calculate the population variance from sample variance

z_score = (samp_mean - pop_mean) / (pop_std / np.sqrt(n))
print('Z-score:',round(z_score,2))


Z-score: 0.08


##### 2. P-value test.

In [261]:
p_value = stats.norm.cdf(z_score)
print('p-value:',round(p_value,2))

p-value: 0.53


##### 3. Confidence Interval.

In [263]:
interval = stats.norm.interval(0.95, loc=samp_mean, scale=pop_std/np.sqrt(n))
print('The confidence interval is:',(interval))

The confidence interval is: (2.932063993209846, 3.0739360067901544)


# Conclusion.

- Here the z-score is greater than -1.64.
- the p-value is greater than 0.05.
- also the confidence interval contains the value in the null hypothesis (i.e. 3). 
- thus we fail to reject (i.e. accept) the null hypothesis. Thus, there is no enough evidence to conclude that on average a typhoid vaccine contains less than 3 mg of ascorbic acid.

##### 1. Z-stastistic test.

In [ ]:
# step1:

# The null and alternative hypothesis is:

# H0:  𝜇≥13 
# H1:  𝜇<13 
# Here ⍺ = 0.05, for a one-tailed test calculate the critical z-value
# left tailed test.

In [264]:
# step2: 

z_val = np.abs(round(stats.norm.isf(q = 0.05), 2))
print('Critical value for one-tailed Z-test:', z_val)

Critical value for one-tailed Z-test: 1.64


In [265]:
# step3:

n = 900
pop_mean = 13
pop_std = 1
samp_mean = 12.5

z_score = (samp_mean - pop_mean) / (pop_std / np.sqrt(n))
print("Z-score:", z_score)

Z-score: -15.0


##### 2. P-value test

In [266]:
# calculate the corresponding p-value for the test statistic
# use 'cdf()' to calculate P(Z <= z_score)
p_value = stats.norm.cdf(z_score)

print('p-value:', p_value)

p-value: 3.6709661993126986e-51


##### 3. Confidence Interval.

In [267]:
# calculate the 95% confidence interval for the population mean
# pass the sample mean to the parameter, 'loc'
# pass the scaling factor (pop_std / n^(1/2)) to the parameter, 'scale'
print('Confidence interval:', stats.norm.interval(0.95, loc = samp_mean, scale = pop_std / np.sqrt(n)))

Confidence interval: (12.434667867181998, 12.565332132818002)


# Conclusion.

- Here the z score is less than -1.64.
- p-value is less than 0.05. 
- also the confidence interval do not contain the value in the null hypothesis (i.e. 13).
- thus we reject the null hypothesis and conclude that the sample is not coming from the population with mean 13.



In [ ]:
# step1:

# The null and alternative hypothesis is:

# H0:  𝜇≤60 
# H1:  𝜇>60 
# Here ⍺ = 0.05, for a one-tailed test calculate the critical z-value.
# Right tailed test.

##### 1. Z-statistic test.

In [269]:
# step2:

z_val = np.abs(round(stats.norm.isf(q = 0.05), 2))
print('Critical value for one-tailed Z-test:', z_val)

Critical value for one-tailed Z-test: 1.64


In [270]:
# step3:

n = 45
pop_mean = 60
pop_std = 30
samp_mean = 75

z_score = (samp_mean - pop_mean) / (pop_std / np.sqrt(n))
print("Z-score:", z_score)

Z-score: 3.3541019662496843


##### 2. P-value test.

In [271]:
# calculate the corresponding p-value for the test statistic
# use 'sf()' to calculate P(Z > z_score)
p_value = stats.norm.sf(z_score)

print('p-value:', p_value)

p-value: 0.0003981150787954058


##### 3. Confidence interval.

In [273]:
print('Confidence interval:', stats.norm.interval(0.95, loc = samp_mean, scale = pop_std / np.sqrt(n)))

Confidence interval: (66.23477459423418, 83.76522540576582)


# Conculsion.

- Here the z score is greater than 1.64.
- the p-value is less than 0.05. 
- also the confidence interval do not contain the value in the null hypothesis (i.e. 60).
- thus we reject the null hypothesis and conclude that the average time to get items delivered is more than 60 minutes.

# ERROR
- The error can occur while choosing the correct hypothesis based on the statistical tests. Two types of errors are **Type I and Type II error**.

#### Type I Error
- This kind of error occurs when we reject the null hypothesis even if it is true
- It is equivalent to a **false positive conclusion**. The probability of type I error is given by the value of  𝛼 , level of significance.

#### Type II Error
- This kind of error occurs when we fail to reject the null hypothesis even if it is wrong,
- It is equivalent to a **false negative conclusion**. The probability of type II error is given by the value of  𝛽 .

# Power of Hypothesis Test.

- The probability of correctly rejecting a false null hypothesis is defined as the Power of a test.
- It is calculated as 1-Beta.

In [ ]:
GRE scores = [311, 313, 309, 307, 302, 313, 305, 336, 322, 300, 335, 312, 282, 314, 310, 316, 301, 315, 299, 298, 308, 
              315, 306, 321, 276, 306, 324, 308, 334, 277, 319, 322, 298, 300, 293, 333,305, 302, 317, 287]

In [8]:
# import the function to calculate the power of test
from statsmodels.stats import power

# given GRE scores
scores = [311, 313, 309, 307, 302, 313, 305, 336, 322, 300, 335, 312, 282, 314, 310, 316, 301, 315, 299, 298, 308, 315, 306, 
          321, 276, 306, 324, 308, 334, 277, 319, 322, 298, 300, 293, 333,305, 302, 317, 287]

# number of observations 
n = len(scores)

# calculate the sample mean
samp_mean = np.mean(scores)

# population mean
pop_mean = 300

# population standard deviation
pop_std = 20


# use 'zt_ind_solve_power()' to calculate the power of the z-test
# keep the value of exact one parameter as 'None' to calculate the value of that parameter using remaining values 
# calculate the 'effect_size' using the sample_mean, null hypothesis value and population standard deviation
# pass the sample size to the parameter, 'nobs1'
# pass the value of alpha to the parameter, 'alpha'
# pass 'None' to the parameter 'power' as we want to calculate its value
# for one sample test, consider the 'ratio' as 0
# pass the alternative for one-tailed test to the parameter, 'alternative'
power = power.zt_ind_solve_power(effect_size = (samp_mean - pop_mean)/pop_std, 
                         nobs1 = n, 
                         alpha = 0.05, 
                         power = None, 
                         ratio = 0, 
                         alternative = 'larger')

print('Power of the test:', power)

Power of the test: 0.8707713725498709


## For finding the Type II error we just do,

- **TYPE II ERROR = (1-Power)**